### PR TITLE
Process slashing with precompute

### PR DESCRIPTION
**Pull request by terencechain**
This PR implements the necessary methods to precompute slashing used for epoch processing. By having total epoch balances precomputed, it drastically increases run time performance

### CODE REVIEW

File: beacon-chain/core/epoch/precompute/slashing_test.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+					{Slashed: true,
+						WithdrawableEpoch: params.BeaconConfig().EpochsPerSlashingsVector / 2,
+						EffectiveBalance:  params.BeaconConfig().MaxEffectiveBalance},
+					{ExitEpoch: params.BeaconConfig().FarFutureEpoch, EffectiveBalance: params.BeaconConfig().MaxEffectiveBalance}},
+				Balances:  []uint64{params.BeaconConfig().MaxEffectiveBalance, params.BeaconConfig().MaxEffectiveBalance},
+				Slashings: []uint64{0, 1e9},
+			},
+			// penalty    = validator balance / increment * (3*total_penalties) / total_balance * increment
+			// 3000000000 = (32 * 1e9)        / (1 * 1e9) * (3*1e9)             / (32*1e9)      * (1 * 1e9)
+			want: uint64(29000000000), // 32 * 1e9 - 3000000000
```
Comment (prestonvanloon) [inline beacon-chain/core/epoch/precompute/slashing_test.go]: Please update the comments. This math doesn't check out

Comment (prestonvanloon) [reply beacon-chain/core/epoch/precompute/slashing_test.go]: edit nvm it does (?) i think

Comment (terencechain) [reply beacon-chain/core/epoch/precompute/slashing_test.go]: Yea the math is correct

File: beacon-chain/core/epoch/precompute/slashing.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/prysmaticlabs/prysm/shared/mathutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// ProcessSlashingsPrecompute processes the slashed validators during epoch processing.
+// This is an optimized version by passing in precomputed total epoch balances.
+func ProcessSlashingsPrecompute(state *pb.BeaconState, p *Balance) *pb.BeaconState {
+	currentEpoch := helpers.CurrentEpoch(state)
+
+	// Compute slashed balances in the current epoch
```
Comment (prestonvanloon) [inline beacon-chain/core/epoch/precompute/slashing.go]: This is a constant? Did you intend to do some computation?

Comment (terencechain) [reply beacon-chain/core/epoch/precompute/slashing.go]: Comment is old. Removed it

**Pull request merged**